noir-projects/noir-contracts/contracts/app/card_game_contract/src/cards.nr

```diff
@@ -78,22 +78,22 @@ impl<Context> Storage<1> for Deck<Context> {
 }
 
 pub fn filter_cards<let N: u32>(
-    retrieved_notes: [Option<RetrievedNote<ValueNote>>; MAX_NOTE_HASH_READ_REQUESTS_PER_CALL],
+    retrieved_notes: BoundedVec<RetrievedNote<ValueNote>, MAX_NOTE_HASH_READ_REQUESTS_PER_CALL>,
     desired_cards: [Card; N],
-) -> [Option<RetrievedNote<ValueNote>>; MAX_NOTE_HASH_READ_REQUESTS_PER_CALL] {
-    let mut selected = [Option::none(); MAX_NOTE_HASH_READ_REQUESTS_PER_CALL];
-
+) -> BoundedVec<RetrievedNote<ValueNote>, MAX_NOTE_HASH_READ_REQUESTS_PER_CALL> {
+    let mut selected = BoundedVec::new();
     let mut found = [false; N];
 
-    for i in 0..retrieved_notes.len() {
-        let retrieved_note = retrieved_notes[i];
-        if retrieved_note.is_some() {
-            let card_note = CardNote::from_note(retrieved_note.unwrap_unchecked().note);
+    for i in 0..MAX_NOTE_HASH_READ_REQUESTS_PER_CALL {
+        if i < retrieved_notes.len() {
```

Comment:
Had to do this because array has len known at comptime but BVec not. So without iterating over MAX_NOTE_HASH_READ_REQUESTS_PER_CALL first it would not compile.

I could use forEach here but that would require me to work with mutable references selected and found and that didn't seem worth it.